web/packages/teleport/src/Discover/Shared/ConnectionDiagnostic/useConnectionDiagnostic.ts

```diff
@@ -20,55 +20,91 @@ import useAttempt from 'shared/hooks/useAttemptNext';
+export function useConnectionDiagnostic() {
   const ctx = useTeleport();
 
-  const { attempt, run } = useAttempt('');
+  const { attempt, setAttempt, handleError } = useAttempt('');
   const [diagnosis, setDiagnosis] = useState<ConnectionDiagnostic>();
   const [ranDiagnosis, setRanDiagnosis] = useState(false);
-  const { emitErrorEvent, emitEvent } = useDiscover();
+  const { emitErrorEvent, emitEvent, prevStep, nextStep, resourceState } =
+    useDiscover();
 
   const access = ctx.storeUser.getConnectionDiagnosticAccess();
   const canTestConnection = access.create && access.edit && access.read;
 
-  function runConnectionDiagnostic(req: ConnectionDiagnosticRequest) {
+  const [showMfaDialog, setShowMfaDialog] = useState(false);
+
+  // runConnectionDiagnostic will initially make a call to check if
+  // resource target requires MFA authentication. After this initial
+  // check depending on if user successfully authenticated or not (
+  // determined by the presence of the token field), will make a call
+  // to test connection.
+  //
+  // Each test connection request will require a MFA check since the
+  // fields for the request may have changed eg: for nodes, the login
+  // field may change.
+  async function runConnectionDiagnostic(
+    req: ConnectionDiagnosticRequest,
+    privilegeTokenId: string
+  ) {
     setDiagnosis(null); // reset since user's can re-test connection.
     setRanDiagnosis(true);
-    run(() =>
-      ctx.agentService
-        .createConnectionDiagnostic(req)
-        .then(diag => {
-          setDiagnosis(diag);
-
-          // The request may succeed, but the connection
-          // test itself can fail:
-          if (!diag.success) {
-            // Append all possible errors:
-            const errors: string[] = [];
-            diag.traces.forEach(trace => {
-              if (trace.status === 'failed') {
-                errors.push(
-                  `[${trace.traceType}] ${trace.error} (${trace.details})`
-                );
-              }
-            });
-            emitErrorEvent(`testing failed: ${errors.join('\n')}`);
-          } else {
-            emitEvent({ stepStatus: DiscoverEventStatus.Success });
+    setShowMfaDialog(false);
+
+    setAttempt({ status: 'processing' });
+
+    try {
+      if (!privilegeTokenId) {
+        const mfaReq = getMfaRequest(req, resourceState);
+        const sessionMfa = await ctx.mfaService.isMfaRequired(mfaReq);
+        if (sessionMfa.required) {
+          setShowMfaDialog(true);
+          return;
+        }
+      }
+
+      const diag = await ctx.agentService.createConnectionDiagnostic({
+        ...req,
+        privilegeTokenId,
+      });
+
+      setAttempt({ status: 'success' });
+      setDiagnosis(diag);
+
+      // The request may succeed, but the connection
+      // test itself can fail:
+      if (!diag.success) {
+        // Append all possible errors:
+        const errors: string[] = [];
```

Comment:
This could be

```suggestion
        const errors = diag.traces
          .filter(trace => trace.status === 'failed')
          .map(trace => `[${trace.traceType}] ${trace.error} (${trace.details})`)
          .join('\n');
```